### PR TITLE
👌 IMPROVE: Add full typing, type checking and test coverage

### PR DESCRIPTION
**Pull request by chrisjsewell**
Hey guys, as the title suggests!

I've also dropped python 3.5 support (as we have also done in aiida-core), added a tox testing configuration, added codecov reporting, and updated the documentation build/ci.

All the tests should pass, except I have left some `# TODO`s regarding some type clarifications and possible errors detected:

```
plumpy/communications.py:29:2: W0511: TODO identifier type? (fixme)
plumpy/communications.py:146:2: W0511: TODO type of task and msg; dict? (fixme)
plumpy/ports.py:306:2: W0511: TODO this was set to None, but that would fail if you tried to compute breadcrumbs (fixme)
plumpy/utils.py:72:2: W0511: TODO I deleted ListenContext and ThreadSafeCounter, (fixme)
plumpy/__init__.py:22:2: W0511: TODO I deleted stack.py, because is not used or tested any where in this code base, or in aiida-core (fixme)
plumpy/process_comms.py:31:2: W0511: TODO process result/status type? (fixme)
plumpy/workchains.py:387:2: W0511: TODO this was added (fixme)
plumpy/process_states.py:253:2: W0511: TODO error: "Kill" has no attribute "result"/"successful" (fixme)
plumpy/base/state_machine.py:250:2: W0511: TODO should class initialise sealed = False? (fixme)
plumpy/base/state_machine.py:257:2: W0511: TODO this appears to never be used? (fixme)
plumpy/persistence.py:646:2: W0511: TODO according to types self._callbacks should first be called, i.e. self._callbacks() (fixme)
```

over to you 😬

### CODE REVIEW

**Review by muhrin — APPROVED** [review]
Thanks Chris!  Looks good to me, let see if the others are happy before merging.

**Comment by chrisjsewell** [comment]
> plumpy/process_states.py:253:2: W0511: TODO error: "Kill" has no attribute "result"/"successful" (fixme)
> plumpy/persistence.py:646:2: W0511: TODO according to types self._callbacks should first be called, i.e. self._callbacks() (fixme)

Both of these are not covered by testing, so are likely to be legitimate bugs (quite possibly related to aiidateam/aiida-core#3888)

File: plumpy/process_comms.py [hunk 1]
```diff
@@ -195,10 +217,15 @@ async def kill_process(self, pid, msg=None):
...
         return result
 
-    async def continue_process(self, pid, tag=None, nowait=False, no_reply=False):
+    async def continue_process(
+        self,
+        pid: 'PID_TYPE',
+        tag: Optional[str] = None,
+        nowait: bool = False,
+        no_reply: bool = False
+    ) -> 'ProcessResult':
```
Comment (unkcpz) [inline plumpy/process_comms.py:228]: if `no_reply=True` method return `None`. So I guess here we have `->Union[None, 'ProcessResult']`

**Review by unkcpz — CHANGES_REQUESTED** [review]
Thanks @chrisjsewell, this really a helpful feature to have type hint. Just some minor request.

Out of curious, when we calling methods with the wrong type, what information will be given to the user? Is type hint only a hint or it really do the type check?

File: plumpy/__init__.py [hunk 2]
```diff
@@ -18,6 +19,8 @@
 from .version import *
 from .workchains import *
 
+# TODO I deleted stack.py, because is not used or tested any where in this code base, or in aiida-core
```
Comment (chrisjsewell) [inline plumpy/__init__.py:22]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -4,37 +4,6 @@ on: [push, pull_request]
...
-                    pip-docs-
-
-        -   name: Set up Python
-            uses: actions/setup-python@v2
-            with:
-                python-version: 3.8
-
-        -   name: Install python dependencies
-            run:
-                pip install -e .[docs,tests]
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply .github/workflows/ci.yml:30]: you are looking at a deleted bit of code here lol.
Documentation CI is now handled directly by ReadTheDocs

File: plumpy/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply plumpy/__init__.py]: removed

File: plumpy/communications.py [hunk 4]
```diff
@@ -17,19 +18,29 @@
...
+    # RPC subscriber params: communicator, msg
+    RpcSubscriber = Callable[[kiwipy.Communicator, Any], Any]
+    # Task subscriber params: communicator, task
+    TaskSubscriber = Callable[[kiwipy.Communicator, Any], Any]
+    # Broadcast subscribers params: communicator, body, sender, subject, correlation id
+    BroadcastSubscriber = Callable[[kiwipy.Communicator, Any, Any, Any, ID_TYPE], Any]
 
-def plum_to_kiwi_future(plum_future):
+
+def plum_to_kiwi_future(plum_future: futures.Future) -> kiwipy.Future:
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply plumpy/communications.py:33]: To clarify, do you mean we should change

```python
if isinstance(result, futures.Future):
```

to

```python
if asyncio.isfuture(result):
```

is the possible source of a bug then?

cc @sphuber

File: plumpy/processes.py [hunk 5]
```diff
@@ -390,52 +409,51 @@ def result(self):
...
+    def successful(self) -> bool:
         """
         Returns whether the result of the process is considered successful
         Will raise if the process is not in the FINISHED state
         """
         try:
-            return self._state.successful
-        except AttributeError as exception:
+            return self._state.successful  # type: ignore
+        except AttributeError  as exception:
```
Comment (chrisjsewell) [inline plumpy/processes.py:425]:
```suggestion
        except AttributeError as exception:
```

File: plumpy/communications.py [hunk 6]
```diff
@@ -88,55 +101,64 @@ def wrap_communicator(communicator, loop=None):
...
         return self._communicator.add_broadcast_subscriber(converted, identifier)
 
-    def remove_broadcast_subscriber(self, identifier):
+    def remove_broadcast_subscriber(self, identifier: 'ID_TYPE') -> None:
         return self._communicator.remove_broadcast_subscriber(identifier)
 
-    def task_send(self, task, no_reply=False):
+    # TODO type of task and msg; dict?
+
+    def task_send(self, task: Any, no_reply: bool = False) -> futures.Future:
```
Comment (chrisjsewell) [inline plumpy/communications.py:143]:
```suggestion
    def task_send(self, task: Any, no_reply: bool = False) -> kiwipy.Future:
```

File: plumpy/communications.py [hunk 7]
```diff
@@ -88,55 +101,64 @@ def wrap_communicator(communicator, loop=None):
...
         return self._communicator.task_send(task, no_reply)
 
-    def rpc_send(self, recipient_id, msg):
+    def rpc_send(self, recipient_id: 'ID_TYPE', msg: Any) -> futures.Future:
```
Comment (chrisjsewell) [inline plumpy/communications.py:146]:
```suggestion
    def rpc_send(self, recipient_id: 'ID_TYPE', msg: Any) -> kiwipy.Future:
```

File: plumpy/process_comms.py (hunk 1 above)
Comment (chrisjsewell) [reply plumpy/process_comms.py:228]:
```suggestion
    ) -> Optional['ProcessResult']:
```

File: plumpy/communications.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply plumpy/communications.py]: fixed

File: plumpy/__init__.py [hunk 8]
```diff
@@ -18,6 +19,7 @@
 from .version import *
 from .workchains import *
 
+
```
Comment (chrisjsewell) [inline plumpy/__init__.py:22]:
```suggestion
```

File: plumpy/communications.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply plumpy/communications.py:33]: @unkcpz what's actualy the difference between `asyncio.Future` and `_asyncio.Future`?

[the remark this answers is not in the document]
Comment (unkcpz) [reply plumpy/communications.py:33]: Emm... That __was__ what I think about and trying to implement in https://github.com/aiidateam/plumpy/pull/183. However, by checking it again, I can not reproduce the issue I mentioned again. The `isinstance` works fine to determine the type of all  the `Future` in the plum.

[the remark this answers is not in the document]
Comment (unkcpz) [reply plumpy/communications.py:33]: For the time being, I think the type hint here is just fine. My bad. I will check this in both plum and aiida_core again.

**Review by unkcpz — DISMISSED** [review]

**Comment by chrisjsewell** [comment]
> Thanks @chrisjsewell, this really a helpful feature to have type hint. Just some minor request.

right this all works now, and requests have been addressed, so I will be merging very shortly; so speak now or for ever hold your peace 😉 

> Out of curious, when we calling methods with the wrong type, what information will be given to the user? Is type hint only a hint or it really do the type check?

type hinting is not used at all at runtime, it is effectively stripped from the code.
But it allows for better auto-completion in editors and for dependant packages to type check against classes/functions used from this package.

Once this is released, I can add proper type hinting/checking to the aiida-core engine code, which I think will be very useful for users/plugin developers writing workflows etc

**Review by unkcpz — APPROVED** [review]
Thanks! all looks good to me.

**Pull request merged**